File: modules/apps/desktop/ui/src/preset/curve.ts
/**
 * The one control of a preset, without drawing it: what the curve costs at
 * each place, where the knob stands, and the settings a place produces.
 *
 * The application works a curve out over the whole range in one pass, so moving
 * the control computes nothing. The line drawn while that answer is on its way
 * is arithmetic over the settings alone, and is shown as an approximation.
 */
import { BOUNDS, DEFAULTS, NOWHERE } from './core'
import type { Curve, Goal, Place, Point, Rule, Settings } from './core'

/** How many places the line drawn in the answer's place is worked out at. */
const PLACES = 25

/** The shortest day a curve of minutes runs to. */
const LEAST_CEILING = 60

/** How far each end of the retention range stands. */
const RETENTION = BOUNDS.retention

/** How long one answer takes where nothing has been answered yet, in seconds. */
const ANSWER = 8

/** Where the saturating line of minutes reaches half of what it ever reaches. */
const HALF = 15

/** The most of the material any day of review brings back. */
const CEILING = 0.98

/** The retention the standing budget is read as buying. */
const MIDDLE = 0.9

/** One setting of a preset a person may take out of the goal's hands. */
export type Field =
  | 'minutesADay'
  | 'newADay'
  | 'reviewsADay'
  | 'retention'
  | 'learned'
  | 'interval'
  | 'byDate'
  | 'counts'
  | 'backlog'
  | 'load'
  | 'evenLoad'

/** Every setting the receipt has a row for, in the order they stand in. */
export const FIELDS: readonly Field[] = [
  'newADay',
  'reviewsADay',
  // The rule for what is learned stands over the value it reads, and one of
  // those values is the target, so the rule stands over that too.
  'learned',
  'interval',
  'retention',
  'minutesADay',
  'byDate',
  'counts',
  'backlog',
  'load',
  'evenLoad',
]

/**
 * The budget each goal schedules by. A goal names one, and the settings of the
 * other two take no part in it: they are neither drawn nor written while it
 * stands, and the file keeps them where the person left them.
 *
 * Minutes are a budget of time. Retention is a budget of cards, since the
 * counts are what close a day worked to a target. A date is neither: the pace
 * follows from the day named, and nothing else may cut it short.
 */
const BUDGETS: Record<Goal, readonly Field[]> = {
  minutes: ['minutesADay'],
  // What a day's budget is spent on belongs to the goal whose budget is cards.
  retention: ['retention', 'newADay', 'reviewsADay', 'counts'],
  date: ['byDate'],
}

/**
 * The share of a day that goes to the debt before anything new is offered. It
 * says what a day is spent on and closes nothing, so it stands beside a budget
 * and not among them. A goal of a date carries the whole material by its own
 * reckoning and has no part in it.
 */
const SPENDING: Record<Goal, readonly Field[]> = {
  minutes: ['backlog'],
  retention: ['backlog'],
  date: [],
}

/**
 * What each rule for the learned reads. The rule itself is drawn under every
 * goal, and under it the one value that rule reads; the other keeps its value
 * and takes no part, as a budget the goal does not name does.
 *
 * A target is read by the goal that steers it and by the rule that counts by
 * it, and either way it is the one key: the receipt draws it once.
 */
const LEARNS: Record<Rule, readonly Field[]> = {
  interval: ['learned', 'interval'],
  retention: ['learned', 'retention'],
}

/** The settings that stand under no goal in particular, and are drawn under all. */
const ALWAYS: readonly Field[] = ['load', 'evenLoad']

/** The settings a goal schedules by, which are the rows the receipt draws. */
export const fieldsUnder = (goal: Goal, learned: Rule): readonly Field[] => {
  const drawn = new Set<Field>([
    ...BUDGETS[goal],
    ...SPENDING[goal],
    ...LEARNS[learned],
    ...ALWAYS,
  ])
  return FIELDS.filter((field) => drawn.has(field))
}

/** The field the goal steers, which is the knob under another name. */
export const steers = (goal: Goal): Field => {
  if (goal === 'minutes') return 'minutesADay'
  if (goal === 'retention') return 'retention'
  return 'byDate'
}

/**
 * The settings that give a curve its shape, as one word. The value the knob
 * rides is among them: the range a curve is drawn over runs to it. A day
 * steers nothing unless it is the goal, so it is left out of the other two:
 * choosing a date and coming back does not make their curves worth asking for
 * again.
 */
export const shapeOf = (settings: Settings): string => {
  const own = new Set<Field>(steers(settings.goal) === 'byDate' ? [] : ['byDate'])
  const said: Record<Field, string> = {
    newADay: `${settings.newADay}`,
    reviewsADay: `${settings.reviewsADay}`,
    retention: `${settings.retention}`,
    minutesADay: `${settings.minutesADay}`,
    byDate: settings.byDate,
    counts: settings.counts,
    backlog: `${settings.backlog}`,
    learned: settings.learned,
    interval: `${settings.interval}`,
    load: Object.keys(settings.load)
      .sort()
      .map((day) => `${day}=${settings.load[day]}`)
      .join(','),
    evenLoad: `${settings.evenLoad}`,
  }
  const rest = FIELDS.filter((field) => !own.has(field)).map((field) => `${field}=${said[field]}`)
  return [settings.goal, ...rest].join(' ')
}

/**
 * What the curve of a goal is read in. A goal of minutes is read in the cards
 * a day answers, which is what a longer day buys; the other two are read in
 * the minutes they cost.
 */
export const costOf = (goal: Goal, point: Point): number =>
  goal === 'minutes' ? point.reviews : point.minutes

/**
 * The budget each goal's own value closes a day by, as the preset writes the
 * key. A target closes no day of its own: what closes a day worked to one is
 * always a count.
 */
const CLOSES: Record<Goal, readonly string[]> = {
  minutes: ['minutes_a_day'],
  retention: [],
  date: ['by_date'],
}

/** Every budget a preset closes a day by, which is the whole of what may be said. */
const CLOSERS: readonly string[] = ['minutes_a_day', 'new_a_day', 'reviews_a_day']

/**
 * What closes the day here besides the goal on screen, and nothing where the
 * goal is the whole of it. A day nothing closed asked for every card there was,
 * and a day two budgets closed names both.
 */
export const limiting = (curve: Curve, point: Point | null): readonly string[] => {
  if (!curve.honest || !point) return []
  return point.closed.filter((one) => CLOSERS.includes(one) && !CLOSES[curve.goal].includes(one))
}

/**
 * The day the overdue pile is gone, read off the very projection the band is
 * drawn from. Null is a place with nothing overdue to be gone at all, and -1
 * is a pile still standing on the last day projected.
 */
export const clearing = (backlog: readonly number[]): number | null => {
  if (!backlog.some((one) => one > 0)) return null
  const at = backlog.indexOf(0)
  return at < 0 ? -1 : at + 1
}

/** A number held inside the bounds of the setting it is. */
export const held = (value: number, of: keyof typeof BOUNDS): number =>
  Math.min(Math.max(value, BOUNDS[of].least), BOUNDS[of].most)

/** The place of the grid nearest a value, and the last one for an empty grid. */
export const nearest = (grid: readonly number[], value: number): number => {
  if (grid.length === 0) return -1
  let at = 0
  for (let i = 1; i < grid.length; i += 1) {
    if (Math.abs((grid[i] ?? 0) - value) < Math.abs((grid[at] ?? 0) - value)) at = i
  }
  return at
}

/** The place a fraction of the way along a grid, held inside it. */
export const placeAt = (grid: readonly number[], share: number): number => {
  if (grid.length === 0) return -1
  const last = grid.length - 1
  return Math.min(Math.max(Math.round(share * last), 0), last)
}

/** The day a number of days from another, written as the year, month and day. */
export const dayAfter = (from: Date, days: number): string => {
  const day = new Date(Date.UTC(from.getUTCFullYear(), from.getUTCMonth(), from.getUTCDate()))
  day.setUTCDate(day.getUTCDate() + days)
  return day.toISOString().slice(0, 10)
}

/** Whether a value is a day at all, which an empty field is not. */
export const isDay = (day: string): boolean => !Number.isNaN(Date.parse(`${day}T00:00:00Z`))

/** How many days stand between today and a day, and zero for a day that is not one. */
export const daysUntil = (today: Date, day: string): number => {
  const named = Date.parse(`${day}T00:00:00Z`)
  if (Number.isNaN(named)) return 0
  const from = Date.UTC(today.getUTCFullYear(), today.getUTCMonth(), today.getUTCDate())
  return Math.round((named - from) / 86400000)
}

/** Why a preset's goal has nothing to work on, and empty where it has. */
export type Idle = 'unpointed' | 'noCards' | 'beginsNothing' | ''

/**
 * Whether the goal has nothing to work on, and why. The counts the curve
 * carries say the first two: no deck points here, or the decks that do hold
 * nothing between them. A preset holding cards is never told it holds none.
 *
 * The third is a preset that schedules and has nothing it can schedule: every
 * card face here is one nobody has begun, and no place of the range begins one.
 * It is a fact about the material, and not a reason the preset is stopped.
 */
export const idle = (curve: Curve): Idle => {
  if (!curve.honest) return ''
  if (curve.decks === 0) return 'unpointed'
  if (curve.cards === 0) return 'noCards'
  if (curve.unbegun === curve.cards && asksNothing(curve)) return 'beginsNothing'
  return ''
}

/** Whether no place of the range asks for a card. A range with no place says nothing. */
const asksNothing = (curve: Curve): boolean =>
  curve.at.length > 0 && curve.at.every((one) => one.reviews === 0)

/**
 * The line the window draws while the application is still working the honest
 * one out. It is arithmetic over the settings alone, and it says so.
 */
export const approximate = (settings: Settings, today: Date): Curve => {
  const grid = gridFor(settings, today)
  const at = grid.map((value) => guessed(settings, value, grid))
  const days = settings.goal === 'date' ? grid.map((value) => dayAfter(today, value)) : []
  const value = standing(settings, today)
  const place = nearest(grid, value)
  const now: Place = { at: place, value, day: days[place] ?? '' }
  return {
    goal: settings.goal,
    grid,
    days,
    at,
    now,
    suggested: NOWHERE,
    decks: 0,
    cards: 0,
    overdue: 0,
    unbegun: 0,
    honest: false,
  }
}

/** Where the preset itself stands, in the units of its goal's grid. */
export const standing = (settings: Settings, today: Date): number => {
  if (settings.goal === 'retention') return settings.retention
  if (settings.goal === 'date') return daysUntil(today, settings.byDate)
  return settings.minutesADay
}

/** The whole range of a goal, at the places the line is drawn at. */
const gridFor = (settings: Settings, today: Date): readonly number[] => {
  if (settings.goal === 'retention') {
    return ladder(RETENTION.least, RETENTION.most, (one) => Math.round(one * 1000) / 1000)
  }
  if (settings.goal === 'date') {
    const most = Math.max(daysUntil(today, settings.byDate), 30)
    return ladder(1, most, Math.round)
  }
  return ladder(0, Math.max(settings.minutesADay, LEAST_CEILING), Math.round)
}

/** The places between two ends, both ends among them. */
const ladder = (least: number, most: number, rounds: (one: number) => number): readonly number[] =>
  Array.from({ length: PLACES }, (_, at) => rounds(least + ((most - least) * at) / (PLACES - 1)))

/**
 * What one place of the range comes to, guessed from the settings. A day of
 * review brings back more of the material the longer it runs, and asking for
 * more of it back costs more of the day.
 */
const guessed = (settings: Settings, value: number, grid: readonly number[]): Point => {
  const flat = {
    retained: 0,
    owed: 0,
    through: 0,
    enough: true,
    closed: [],
    clears: 0,
    learned: 0,
    short: 0,
    backlog: [],
  }
  if (settings.goal === 'retention') {
    const minutes = (settings.minutesADay || DEFAULTS.minutesADay) * ((1 - MIDDLE) / (1 - value))
    return { ...flat, minutes, reviews: (minutes * 60) / ANSWER, retained: value }
  }
  if (settings.goal === 'date') {
    const span = Math.max(grid[grid.length - 1] ?? 1, 1)
    const minutes = ((settings.minutesADay || DEFAULTS.minutesADay) * span) / Math.max(value, 1)
    const enough = minutes <= (settings.minutesADay || DEFAULTS.minutesADay)
    const through = Math.min(value / span, 1)
    return { ...flat, minutes, reviews: (minutes * 60) / ANSWER, through, enough }
  }
  const retained = (CEILING * value) / (value + HALF)
  return { ...flat, minutes: value, reviews: (value * 60) / ANSWER, retained }
}

/**
 * The settings one place of the curve produces, which is the goal's own value
 * off the grid. Every other setting stands as the person left it.
 */
export const producing = (
  was: Settings,
  place: number,
  curve: Curve,
  today: Date,
): Settings => {
  const value = curve.grid[place] ?? standing(was, today)
  if (curve.goal === 'retention') {
    return { ...was, retention: held(round(value, 2), 'retention') }
  }
  if (curve.goal === 'date') {
    return { ...was, byDate: curve.days[place] ?? dayAfter(today, value) }
  }
  return { ...was, minutesADay: held(Math.round(value), 'minutesADay') }
}

/** A number to that many places. */
export const round = (value: number, places: number): number => {
  const scale = 10 ** places
  return Math.round(value * scale) / scale
}

File: modules/libs/core/flashcards/review/curve.go
package review

import (
	"context"
	"math"
	"time"
)

// Points is how many places a curve is worked out at. A goal of minutes and a
// goal of retention run at exactly this many, and a goal of a date whose range
// is shorter runs at one place a day.
const Points = 25

// MostAhead is how far ahead a goal of a date is projected. A day further off
// than this stands nowhere on the range, and the curve is drawn to the end of
// the range.
const MostAhead = 5 * 365

// Curve is what the one control of a preset comes to over the whole range of
// its goal. The grid, what stands at each place of it, and the two places
// pointed at are all here.
type Curve struct {
	Goal Goal
	// Grid is the value of the goal at each place: minutes for minutes_a_day, a
	// share of cards for retention, and days from today for by_date.
	Grid []float64
	// Days names the day of each place, and is filled for a goal of a date.
	Days []string
	// Points are what the preset comes to at each place of the grid.
	Points []Point
	// Now is where the preset stands.
	Now Place
	// Suggested is the place worth pointing at: under a goal of minutes the
	// shortest day the clock no longer cuts short, and under a goal of a date
	// the soonest day the material is learned by at a cost the minutes the
	// preset keeps allow. A goal of retention has none, and stands at Nowhere.
	Suggested Place
	// Stops is why the settings this curve was drawn under schedule nothing,
	// and is empty where they schedule something. It is asked of the settings
	// the request carried.
	Stops StopReason
	// Decks is how many decks are scheduled by this preset. Zero is a preset no
	// deck points at, and every place of the curve stands at zero with it.
	Decks int
	// Cards is how many card faces stand in those decks. Zero is a preset with
	// nothing to schedule, and every place of the curve stands at zero with it.
	Cards int
	// Overdue is how many of those card faces have had their day and were not
	// answered on it, and Unbegun how many nobody has answered at all, so they
	// have had no day. Both are one number over the whole curve: facts about the
	// vault as it stands, and not about the setting being chosen.
	Overdue int
	Unbegun int
}

// Point is what a preset comes to at one place of the grid.
//
// Reviews is counted in card faces and Minutes in the showings they take, so a
// face the day comes back to costs its minutes and is the one card.
//
// What the two are the height of is the goal's own question. Under a goal of
// minutes they are the first day of the run the preset admits: the next session
// a person will actually sit down to. Under a goal of retention they are the
// load over the days the preset admits.
//
// A goal of a date reads Reviews off that first day, and fills Minutes with
// what getting through the material by the day the place names costs, over the
// days of review up to it.
type Point struct {
	Reviews float64
	Minutes float64
	// Retained is the share of the material that comes back on the last day
	// projected, which is what the load beside it buys.
	Retained float64
	// Owed is the card faces standing owed on the last day the projection ran:
	// Ahead days off under a goal of minutes or of retention, and the day the
	// place names under a goal of a date.
	Owed int
	// Share is the share of the material learned by this day, and Enough is
	// whether the pace this place sets learns every card face that can be
	// learned by it. A goal keeping no such account stands at true.
	Share  float64
	Enough bool
	// Short is how many card faces cannot be learned by this day whatever the
	// pace, which is the rule wanting more days than the day leaves them.
	Short int
	// Closed is every budget that closed the day here, in the words the preset
	// writes them in, and is empty where the material itself ran out.
	Closed BudgetNames
	// Clears is how many days of review at this place it takes before nothing
	// is overdue. A curve standing over nothing overdue clears in none, and a
	// place whose pace never gets there is NeverClears.
	Clears int
	// Learned is how many card faces stand learned today at this place, and
	// Learns how many days of review it takes before all of them do. A place
	// whose horizon ends with one still to learn is NeverLearns, and one with no
	// such day to name is LearnsUnasked.
	Learned int
	Learns  int
	// Backlog is how many card faces stand overdue at the end of each day
	// projected at this place, one entry a day over the whole horizon. It runs
	// over days and not over the goal's range.
	Backlog []int
}

// Place is one place on the curve worth pointing at.
type Place struct {
	// Index is the place of the grid, and is -1 when the value falls outside it.
	Index int
	// Value is the goal's value at the place, in the units of the grid. What the
	// preset stands at need not sit on the grid.
	Value float64
	// Day is the day at the place, and is filled for a goal of a date.
	Day string
}

// Nowhere is a place that falls outside the grid.
var Nowhere = Place{Index: -1}

// drawing is one curve being worked out: the run every place of it is projected
// with, and what stands the same at every place.
type drawing struct {
	run    Simulation
	now    time.Time
	preset Preset
	at     map[CardFaceID]Schedule
	unseen int
	by     func(retention float64) Scheduler
	places func(count int, each func(at int) error) error
}

// Curve is what the one control of a preset comes to over the whole range of
// its goal, worked out by running this projection at every place of the range.
//
// The map is where the answers have left every card face the preset schedules,
// and unseen is how many of its card faces nobody has answered, as Run reads
// them. By is the scheduler asking for a share of the cards to come back, and
// is asked again at each place of a goal of retention.
//
// Places runs the places of the range and is where a caller says how many of
// them run at once; a caller handing over none runs them one at a time.
func (s Simulation) Curve(
	ctx context.Context, now time.Time, p Preset, at map[CardFaceID]Schedule,
	unseen int, by func(retention float64) Scheduler,
	places func(count int, each func(at int) error) error,
) (Curve, error) {
	if by == nil {
		by = func(retention float64) Scheduler { return NewFSRSAt(retention) }
	}
	if places == nil {
		places = func(count int, each func(at int) error) error {
			for i := range count {
				if err := each(i); err != nil {
					return err
				}
			}
			return nil
		}
	}
	d := drawing{run: s, now: now, preset: p, at: at, unseen: unseen, by: by, places: places}
	switch p.Goal {
	case GoalRetention:
		return d.retention(ctx)
	case GoalDate:
		return d.date(ctx)
	default:
		return d.minutes(ctx)
	}
}

// minutes is the curve of how long a day of review runs.
//
// It runs from a short day to twice what carrying the whole load costs, so the
// place where the load is carried stands inside it. Each place is the session a
// person would sit down to now, which is the day the deck screen offers.
func (d drawing) minutes(ctx context.Context) (Curve, error) {
	run, now, p, at, unseen := d.run, d.now, d.preset, d.at, d.unseen

	free := p
	free.MinutesADay = int(MinutesADayBounds.Most)
	load, err := run.Run(ctx, now, free, at, unseen)
	if err != nil {
		return Curve{}, err
	}

	out := Curve{Goal: GoalMinutes, Now: Nowhere, Suggested: Nowhere}
	top := ceiling(carried(load), float64(p.MinutesADay))
	for i := range Points {
		out.Grid = append(out.Grid, math.Round(top*float64(i+1)/Points))
	}
	// The day the preset keeps is a place of the grid.
	snap(out.Grid, float64(p.MinutesADay), 0, len(out.Grid)-1)
	// A place is read on the last day of its run, and that is the day the run
	// works the returning share out on.
	run.Retains = []int{run.Covers() - 1}
	out.Points = make([]Point, len(out.Grid))
	if err := d.places(len(out.Grid), func(i int) error {
		one := p
		one.MinutesADay = int(out.Grid[i])
		ran, err := run.Run(ctx, now, one, at, unseen)
		if err != nil {
			return err
		}
		place := session(ran)
		if place.Learns, err = learnt(ctx, run, now, one, at, unseen); err != nil {
			return err
		}
		out.Points[i] = place
		return nil
	}); err != nil {
		return Curve{}, err
	}

	out.Now = Place{Index: nearest(out.Grid, float64(p.MinutesADay)), Value: float64(p.MinutesADay)}
	// What is suggested is the shortest day that asks everything the day holds:
	// the minutes stop closing it, and the material is what runs out. A load
	// nothing on the grid carries is suggested at the longest day on it.
	out.Suggested = Place{Index: len(out.Grid) - 1, Value: out.Grid[len(out.Grid)-1]}
	for i, one := range out.Points {
		if len(one.Closed) == 0 {
			out.Suggested = Place{Index: i, Value: out.Grid[i]}
			break
		}
	}
	return out, nil
}

// retention is the curve of how much is asked of memory.
//
// A higher target is shorter intervals and more reviews, so under a budget that
// binds it is also fewer cards kept up with.
//
// Each place is the load over the days projected, which is what the target costs
// week after week. Today is no part of it: the schedules a day opens with are
// the same whatever target is chosen, and what a target changes it changes from
// tomorrow on.
func (d drawing) retention(ctx context.Context) (Curve, error) {
	run, now, p, at, unseen := d.run, d.now, d.preset, d.at, d.unseen

	out := Curve{Goal: GoalRetention, Now: Nowhere, Suggested: Nowhere}
	least, most := RetentionBounds.Least, RetentionBounds.Most
	for i := range Points {
		out.Grid = append(out.Grid, least+(most-least)*float64(i)/float64(Points-1))
	}
	// The target the preset asks for is a place of the grid. The two ends are
	// what memory allows, and no setting displaces them.
	snap(out.Grid, p.Retention, 1, len(out.Grid)-2)

	// A place is read on the last day of its run, and that is the day the run
	// works the returning share out on.
	run.Retains = []int{run.Covers() - 1}
	out.Points = make([]Point, len(out.Grid))
	if err := d.places(len(out.Grid), func(i int) error {
		one, asks := p, run
		one.Retention = out.Grid[i]
		asks.By = d.by(out.Grid[i])
		ran, err := asks.Run(ctx, now, one, at, unseen)
		if err != nil {
			return err
		}
		place := point(ran)
		if place.Learns, err = learnt(ctx, asks, now, one, at, unseen); err != nil {
			return err
		}
		out.Points[i] = place
		return nil
	}); err != nil {
		return Curve{}, err
	}

	out.Now = Place{Index: nearest(out.Grid, p.Retention), Value: p.Retention}
	// A goal of retention suggests nothing, and its place stands at Nowhere.
	return out, nil
}

// date is the curve of getting through the material by a day.
//
// Each day of it carries what a day of review has to run to be through by then,
// and what the budget the preset keeps gets through by then.
//
// The range stands on the vault and not on the day the file names, so a person
// can always give themselves longer than they have. It begins tomorrow, because
// a date of today is no period at all, and reaches whichever is further off:
// twice as far as the day named, or the day the material would be through at
// one card a day, which is the slowest a day of review goes.
func (d drawing) date(ctx context.Context) (Curve, error) {
	run, now, p, at, unseen := d.run, d.now, d.preset, d.at, d.unseen

	out := Curve{Goal: GoalDate, Now: Nowhere, Suggested: Nowhere}
	open := run.Day.Opens(now)
	by := p.By.Format(Named)
	if p.By.IsZero() || by < run.Day.Names(open) {
		return out, nil
	}
	// How far off the day the file names is, counting the day holding now as
	// none. A day further off than the projection reaches stands nowhere on the
	// range, and the range is drawn as far as it goes.
	named := 0
	for day := open; run.Day.Names(day) < by; day = day.AddDate(0, 0, 1) {
		named++
		if named > MostAhead {
			named = Nowhere.Index
			break
		}
	}

	first := 1
	last := min(MostAhead, max(2*named, unseen, first))
	if named < 0 {
		last = MostAhead
	}
	run.Days = last + 1

	// Each place of the range is one day named, run at the pace that day sets:
	// the material spread over the days up to it. Everything read off a place
	// comes from that one run.
	//
	// A place runs a horizon of Ahead days, and of its own day where that stands
	// further off.
	steps := naming(spread(last-first+1, Points), named-first)
	out.Grid = make([]float64, len(steps))
	out.Days = make([]string, len(steps))
	out.Points = make([]Point, len(steps))
	if err := d.places(len(steps), func(i int) error {
		day := first + steps[i]
		aiming, asks := p, run
		aiming.By = open.AddDate(0, 0, day)
		asks.Days = max(day+1, Ahead)
		// A place of this range is read on the day it names, and that is the day
		// the run works the returning share out on.
		asks.Retains = []int{day}
		ran, err := asks.Run(ctx, now, aiming, at, unseen)
		if err != nil {
			return err
		}
		back, _ := ran.Retained.On(day)
		// A day at none of the load is no session at all, so what a day of
		// review holds is read off the first day this run admits.
		opening, session := ran.Session()
		out.Grid[i] = float64(day)
		out.Days[i] = run.Day.Names(aiming.By)
		one := Point{
			// What it costs is what the days up to that one spend, and the days
			// past it are no part of getting through by it.
			Minutes: costing(ran.Spent[:day+1], ran.Admitted[:day+1]),
			// A place of this range is read on the day it names. Past that day
			// the preset schedules nothing, so a debt read off the end of the
			// horizon is a debt nobody was asked to pay.
			Owed:     ran.Backlog[day],
			Retained: back,
			Share:    ran.Through[day],
			Enough:   reached(ran, day, ran.Short),
			Short:    ran.Short,
			Closed:   BudgetNames{ClosedPaused},
			Clears:   ran.Clears,
			Learned:  ran.Learned,
			Learns:   ran.Learns,
			Backlog:  ran.Backlog,
		}
		if session {
			one.Reviews, one.Closed = float64(ran.Faced[opening]), ran.Closed[opening]
		}
		out.Points[i] = one
		return nil
	}); err != nil {
		return Curve{}, err
	}

	// The day the file names is a place of the grid, so what stands under the
	// place is worked out for that day.
	if named >= 0 {
		out.Now = Place{
			Index: nearest(out.Grid, float64(named)),
			Value: float64(named),
			Day:   run.Day.Names(open.AddDate(0, 0, named)),
		}
	}
	// What is suggested is the soonest day the material can be learned by:
	// nothing out of reach on it and the pace through the whole of it. Where the
	// preset keeps minutes, it is the soonest such day whose cost fits them.
	if p.MinutesADay > 0 {
		for i, one := range out.Points {
			if learns(one) && one.Minutes <= float64(p.MinutesADay) {
				out.Suggested = Place{Index: i, Value: out.Grid[i], Day: out.Days[i]}
				return out, nil
			}
		}
	}
	// A preset keeping no minutes, and a range no day of which fits them, are
	// suggested the soonest day that gets there at whatever it costs. A range no
	// day of which gets there points at none.
	for i, one := range out.Points {
		if learns(one) {
			out.Suggested = Place{Index: i, Value: out.Grid[i], Day: out.Days[i]}
			return out, nil
		}
	}
	return out, nil
}

// learnt is how many days of review it takes before the whole material stands
// learned, projected on the assumption that nothing is forgotten.
//
// The day it names is the day a person reaches if the answers go well. It is
// the one figure drawn under that assumption, and the run and the arithmetic
// are those of every figure beside it.
func learnt(
	ctx context.Context, run Simulation, now time.Time, p Preset,
	at map[CardFaceID]Schedule, unseen int,
) (int, error) {
	run.Recalls = NothingForgotten
	// The day the material is learned is all this run is read for.
	run.Retains = nil
	ran, err := run.Run(ctx, now, p, at, unseen)
	if err != nil {
		return 0, err
	}
	return ran.Learns, nil
}

// learns reports whether the whole material stands learned on this day: no card
// face out of reach of it, and the pace it sets through every one of them.
func learns(one Point) bool { return one.Short == 0 && one.Enough }

// reached reports whether every card face that can be learned by this day of a
// run stands learned on it. Short is how many cannot be, whatever the pace.
func reached(p Projection, day, short int) bool {
	if p.Faces == 0 {
		return true
	}
	return p.Through[day] >= float64(p.Faces-short)/float64(p.Faces)
}

// point is a projection as one place of a curve, at the load it carries over
// the days the preset admits.
func point(p Projection) Point {
	// A place is read on the last day of its run, and that is the day the run
	// works the returning share out on.
	back, _ := p.Retained.On(p.Days - 1)
	return Point{
		Reviews: p.ReviewsADay,
		Minutes: p.MinutesADay,
		// A goal of minutes and a goal of retention set no pace at a day, so no
		// place of theirs falls short of one.
		Enough:   true,
		Retained: back,
		Owed:     p.Owed,
		Share:    p.Through[len(p.Through)-1],
		Closed:   closing(p),
		Short:    p.Short,
		Clears:   p.Clears,
		Learned:  p.Learned,
		Learns:   p.Learns,
		Backlog:  p.Backlog,
	}
}

// closing is what closed the first day the preset admits. A preset admession no
// day is closed by the pause.
func closing(p Projection) BudgetNames {
	day, any := p.Session()
	if !any {
		return BudgetNames{ClosedPaused}
	}
	return p.Closed[day]
}

// session is a projection as one place of a curve, at the first day of it the
// preset admits: the next session a person will sit down to.
//
// It is one real day of the run, worked out by the arithmetic the deck screen
// runs, so the count here is the count that session hands a person. A preset
// admession no day at all holds no session, and stands at nothing.
func session(p Projection) Point {
	out := point(p)
	day, any := p.Session()
	if !any {
		return out
	}
	out.Reviews, out.Minutes = float64(p.Faced[day]), p.Spent[day].Minutes()
	return out
}

// carried is how long the first day the preset admits took, which is what
// carrying the whole load costs on the next session.
func carried(p Projection) float64 {
	day, any := p.Session()
	if !any {
		return 0
	}
	return p.Spent[day].Minutes()
}

// costing is how long a day of review runs over these days, in minutes.
//
// A day the preset does not admit is no session at all and takes no part: a
// week of five days runs its minutes over five days.
func costing(spent []time.Duration, admitted []bool) float64 {
	var all time.Duration
	days := 0
	for i, one := range spent {
		if i < len(admitted) && !admitted[i] {
			continue
		}
		all += one
		days++
	}
	if days == 0 {
		return 0
	}
	return all.Minutes() / float64(days)
}
